consensus/src/network_interface.rs

```diff
@@ -139,9 +170,33 @@ impl ConsensusNetworkSender {
         message: ConsensusMsg,
         timeout: Duration,
     ) -> Result<ConsensusMsg, RpcError> {
-        let protocol = ProtocolId::ConsensusRpc;
-        self.network_sender
-            .send_rpc(recipient, protocol, message, timeout)
-            .await
+        let remote_protocols = self.connection_metadata(recipient)?;
+        for protocol in RPC {
+            if remote_protocols.contains(*protocol) {
+                return self
+                    .network_sender
+                    .send_rpc(recipient, *protocol, message, timeout)
+                    .await;
+            }
+        }
+        Err(anyhow!("No available protocols for peer {}", recipient).into())
+    }
+
+    /// Initialize a shared hashmap about connections metadata that is updated by the receiver.
+    pub fn initialize(&mut self, connections: Arc<RwLock<HashMap<PeerId, SupportedProtocols>>>) {
+        self.connections = Some(connections);
+    }
+
+    /// Query the connection metadata from the peer.
+    fn connection_metadata(&self, peer: PeerId) -> anyhow::Result<SupportedProtocols> {
```

Comment:
This is misleading since it doesn't return the `ConnectionMetadata`.  I'd rename it `supported_protocols`.  You could also split this out into `DirectSend` and `Rpc` so you wouldn't have to iterate through the intersection in the code above.